internal/identity/gitlab.go

```diff
@@ -0,0 +1,118 @@
+package identity // import "github.com/pomerium/pomerium/internal/identity"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	oidc "github.com/pomerium/go-oidc"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/version"
+	"golang.org/x/oauth2"
+)
+
+const defaultGitLabProviderURL = "https://gitlab.com"
+const defaultGitLabGroupURL = "https://gitlab.com/api/v4/groups"
+const revokeURL = "https://gitlab.com/oauth/revoke"
+
+// GitLabProvider is an implementation of the OAuth Provider
+type GitLabProvider struct {
+	*Provider
+
+	RevokeURL string `json:"revocation_endpoint"`
+}
+
+// NewGitLabProvider returns a new GitLabProvider.
+// https://www.pomerium.io/docs/identity-providers.html#gitlab-active-directory
+func NewGitLabProvider(p *Provider) (*GitLabProvider, error) {
+	ctx := context.Background()
+
+	if p.ProviderURL == "" {
+		p.ProviderURL = defaultGitLabProviderURL
+	}
+
+	var err error
+	p.provider, err = oidc.NewProvider(ctx, p.ProviderURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(p.Scopes) == 0 {
+		p.Scopes = []string{oidc.ScopeOpenID, "api", "read_user", "profile", "email"}
+	}
+
+	p.verifier = p.provider.Verifier(&oidc.Config{ClientID: p.ClientID})
+	p.oauth = &oauth2.Config{
+		ClientID:     p.ClientID,
+		ClientSecret: p.ClientSecret,
+		Endpoint:     p.provider.Endpoint(),
+		RedirectURL:  p.RedirectURL.String(),
+		Scopes:       p.Scopes,
+	}
+	gp := &GitLabProvider{
+		Provider:  p,
+		RevokeURL: revokeURL,
+	}
+
+	if err := p.provider.Claims(&gp); err != nil {
+		return nil, err
+	}
+
+	p.UserGroupFn = gp.UserGroups
+
+	return gp, nil
+}
+
+// Revoke attempts to revoke session access via revocation endpoint
+// https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#revoking-a-personal-access-token
+func (p *GitLabProvider) Revoke(ctx context.Context, token *oauth2.Token) error {
+	params := url.Values{}
+	params.Add("access_token", token.AccessToken)
+
+	err := httputil.Client(ctx, http.MethodPost, p.RevokeURL, version.UserAgent(), nil, params, nil)
+	if err != nil && err != httputil.ErrTokenRevoked {
+		return err
+	}
+
+	return nil
+}
+
+// GetSignInURL returns the sign in URL with typical oauth parameters
+func (p *GitLabProvider) GetSignInURL(state string) string {
+	return p.oauth.AuthCodeURL(state, oauth2.AccessTypeOffline, oauth2.SetAuthURLParam("prompt", "select_account"))
```

Comment:
Does this require the additional params for  `prompt select_account`? I wasn't able to get auth flow working without removing this.

Reply:
@desimone I'll love to have a call with you to sort all these out.

It will be a good idea to have the update of the doc in a separate PR, what do you think?

Reply:
I have added the docs for gitlab support